smart_open/smart_open_lib.py

```diff
@@ -604,15 +602,136 @@ def file_smart_open(fname, mode='rb'):
+
+        If none of those are set, will connect unauthenticated.
+        """
+        PY2 = sys.version_info[0] == 2
+        if PY2:
+            from urllib2 import urlopen
+        else:
+            from urllib.request import urlopen
+
+        if kerberos:
+            import requests_kerberos
+            auth = requests_kerberos.HTTPKerberosAuth()
+        elif user is not None and password is not None:
+            auth = (user, password)
+        else:
+            auth = None
+        
+        self.response = requests.get(url, auth=auth, stream=True)
+
+        if not self.response.ok:
+            self.response.raise_for_status()
+
+        self._read_buffer = None
+        self._read_iter = None
+        self._readline_iter = None
+
+    def __iter__(self):
+        return response.iter_lines()
+
+    def readline(self):
+        """
+        Mimics the readline call to a filehandle object.
+        """
+        if self._readline_iter is None:
+            self._readline_iter = self.response.iter_lines()
+
+        try:
+            return next(self._readline_iter)
+        except StopIteration:
+            raise EOFError()
+
+    def readlines(self):
+        """
+        Mimics the readlines call to a filehandle object.
+        """
+        return list(self.response.iter_lines())
+
+    def read(self, size=None):
+        """
+        Mimics the read call to a filehandle object.
+        """
+        if size is None:
+            return self.response.content
+        else:
+            if self._read_iter is None:
+                self._read_iter = self.response.iter_content(size)
+                self._read_buffer = next(self._read_iter)
+            
+            while len(self._read_buffer) < size:
+                try:
+                    self._read_buffer += next(self._read_iter)
+                except StopIteration:
+                    # Oops, ran out of data early.
+                    retval = self._read_buffer
+                    self._read_buffer = ''
+                    if len(retval) == 0:
+                        raise EOFError()
+                    else:
+                        return retval
+            
+            # If we got here, it means we have enough data in the buffer
+            # to return to the caller.
+            retval = self._read_buffer[:size]
+            self._read_buffer = self._read_buffer[size:]
+            return retval
+
+    def __enter__(self, *args, **kwargs):
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        self.response.close()
+
+def HttpOpenRead(parsed_uri, mode='r', **kwargs):
+    if parsed_uri.scheme not in ('http', 'https'):
+        raise TypeError("can only process http/https urls")
+    if mode not in ('r', 'rb'):
+        raise NotImplementedError('Streaming write to http not supported')
+
+    url = parsed_uri.uri_path
+
+    response = HttpReadStream(url, **kwargs)
+
+    if url.endswith('.gz'):
```

Comment:
the gz and bz2 handling code is identical to `file_smart_open`. Please refactor to avoid duplication.